In [2]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import yfinance as yf
import sys
sys.path.append("../utils")

from indicators import compute_rsi, compute_macd, compute_bbands

print("All imports successful")

All imports successful


In [6]:
def full_analysis(ticker, start="2019-01-01", end="2024-01-01"):
    """
    Complete Week 1 analysis pipeline for my ticker.
    Returns summary stats dict.
    """
    print(f"\n{'#'*50}")
    print(f" FULL ANALYSIS: {ticker}")
    print(f"{'#'*50}")

    # ---Data---
    df=yf.download(ticker, start=start, end=end, auto_adjust=True)
    df.columns=df.columns.get_level_values(0)
    price=df["Close"].squeeze()
    returns=np.log(price/price.shift(1)).dropna()

    # ---Core stats---
    ann_ret=returns.mean()*252
    ann_vol=returns.std()*np.sqrt(252)
    sharpe=(ann_ret-0.02)/ann_vol

    # ---Max drawdown---
    wealth=(1+returns).cumprod()
    drawdown=(wealth-wealth.cummax())/wealth.cummax()
    max_dd=drawdown.min()

    # ---Distribution---
    skew=returns.skew()
    kurt=returns.kurt()

    # --- VaR ---
    var_95 = returns.sort_values().iloc[int(0.05 * len(returns))]
    cvar_95 = returns.sort_values().iloc[int(0.05 * len(returns))].mean()

    # ---Current indicators ---
    rsi_val=compute_rsi(price).iloc[-1]
    macd_val, sig_val, _=compute_macd(price)
    macd_signal="BULLISH" if macd_val.iloc[-1]> sig_val.iloc[-1] else "BEARISH"

    # ---Print summary ---
    print(f"\n  PERFORMANCE")
    print(f"  Ann. Return    : {ann_ret:>8.2%}")
    print(f"  Ann. Volatility: {ann_vol:>8.2%}")
    print(f"  Sharpe Ratio   : {sharpe:>8.2f}")
    print(f"  Max Drawdown   : {max_dd:>8.2%}")
    print(f"\n  DISTRIBUTION")
    print(f"  Skewness       : {skew:>8.4f}")
    print(f"  Excess Kurtosis: {kurt:>8.4f}")
    print(f"  95% VaR        : {var_95:>8.2%}")
    print(f"  95% CVaR       : {cvar_95:>8.2%}")
    print(f"\n  CURRENT SIGNALS")
    print(f"  RSI(14)        : {rsi_val:>8.2f}")
    print(f"  MACD Signal    : {macd_signal:>8}")

    return {
        "ticker"  : ticker,
        "ann_ret" : ann_ret,
        "ann_vol" : ann_vol,
        "sharpe"  : sharpe,
        "max_dd"  : max_dd,
        "skew"    : skew,
        "kurt"    : kurt,
        "var_95"  : var_95,
        "rsi"     : rsi_val,
        "macd"    : macd_signal
    }

In [7]:
spy_result=full_analysis("SPY")


##################################################
 FULL ANALYSIS: SPY
##################################################


[*********************100%***********************]  1 of 1 completed


  PERFORMANCE
  Ann. Return    :   14.50%
  Ann. Volatility:   21.08%
  Sharpe Ratio   :     0.59
  Max Drawdown   :  -35.75%

  DISTRIBUTION
  Skewness       :  -0.8213
  Excess Kurtosis:  12.4175
  95% VaR        :   -1.96%
  95% CVaR       :   -1.96%

  CURRENT SIGNALS
  RSI(14)        :    70.54
  MACD Signal    :  BEARISH


In [8]:
tickers = ["SPY", "AAPL", "BTC-USD", "GLD", "TLT"]
results= []

for t in tickers:
    r = full_analysis(t)
    results.append(r)


##################################################
 FULL ANALYSIS: SPY
##################################################


[*********************100%***********************]  1 of 1 completed



  PERFORMANCE
  Ann. Return    :   14.50%
  Ann. Volatility:   21.08%
  Sharpe Ratio   :     0.59
  Max Drawdown   :  -35.75%

  DISTRIBUTION
  Skewness       :  -0.8213
  Excess Kurtosis:  12.4175
  95% VaR        :   -1.96%
  95% CVaR       :   -1.96%

  CURRENT SIGNALS
  RSI(14)        :    70.54
  MACD Signal    :  BEARISH

##################################################
 FULL ANALYSIS: AAPL
##################################################


[*********************100%***********************]  1 of 1 completed



  PERFORMANCE
  Ann. Return    :   32.59%
  Ann. Volatility:   32.25%
  Sharpe Ratio   :     0.95
  Max Drawdown   :  -35.18%

  DISTRIBUTION
  Skewness       :  -0.2671
  Excess Kurtosis:   5.4428
  95% VaR        :   -3.12%
  95% CVaR       :   -3.12%

  CURRENT SIGNALS
  RSI(14)        :    51.12
  MACD Signal    :  BEARISH

##################################################
 FULL ANALYSIS: BTC-USD
##################################################


[*********************100%***********************]  1 of 1 completed



  PERFORMANCE
  Ann. Return    :   33.11%
  Ann. Volatility:   56.25%
  Sharpe Ratio   :     0.55
  Max Drawdown   :  -83.72%

  DISTRIBUTION
  Skewness       :  -1.2283
  Excess Kurtosis:  19.0454
  95% VaR        :   -5.15%
  95% CVaR       :   -5.15%

  CURRENT SIGNALS
  RSI(14)        :    51.09
  MACD Signal    :  BEARISH

##################################################
 FULL ANALYSIS: GLD
##################################################


[*********************100%***********************]  1 of 1 completed



  PERFORMANCE
  Ann. Return    :    9.11%
  Ann. Volatility:   14.95%
  Sharpe Ratio   :     0.48
  Max Drawdown   :  -23.89%

  DISTRIBUTION
  Skewness       :  -0.2973
  Excess Kurtosis:   3.0730
  95% VaR        :   -1.58%
  95% CVaR       :   -1.58%

  CURRENT SIGNALS
  RSI(14)        :    59.85
  MACD Signal    :  BULLISH

##################################################
 FULL ANALYSIS: TLT
##################################################


[*********************100%***********************]  1 of 1 completed


  PERFORMANCE
  Ann. Return    :   -2.02%
  Ann. Volatility:   17.63%
  Sharpe Ratio   :    -0.23
  Max Drawdown   :  -51.11%

  DISTRIBUTION
  Skewness       :   0.1172
  Excess Kurtosis:   4.4238
  95% VaR        :   -1.69%
  95% CVaR       :   -1.69%

  CURRENT SIGNALS
  RSI(14)        :    63.37
  MACD Signal    :  BEARISH


In [9]:
df_results = pd.DataFrame(results)
df_results = df_results.set_index("ticker")
df_ranked = df_results.sort_values("sharpe", ascending=False)

print(f"\n{'='*65}")
print(f" 5-ASSET COMPARISON - RANKED BY SHARPE")
print(f"{'='*65}")
print(f"{'Asset':<12} {'Ann Ret':>9} {'Ann Vol':>9} {'Sharpe':>8} {'Max DD':>9} {'Kurt':>7}")
print("-"* 65)
for idx, row in df_ranked.iterrows():
    print(f"{idx:<12} {row['ann_ret']:>9.2%} {row['ann_vol']:>9.2%} {row['sharpe']:>8.2f} {row['max_dd']:>9.2%} {row['kurt']:>7.2f}")
print("=" *65)


 5-ASSET COMPARISON - RANKED BY SHARPE
Asset          Ann Ret   Ann Vol   Sharpe    Max DD    Kurt
-----------------------------------------------------------------
AAPL            32.59%    32.25%     0.95   -35.18%    5.44
SPY             14.50%    21.08%     0.59   -35.75%   12.42
BTC-USD         33.11%    56.25%     0.55   -83.72%   19.05
GLD              9.11%    14.95%     0.48   -23.89%    3.07
TLT             -2.02%    17.63%    -0.23   -51.11%    4.42


# Week 1 Investment Memo - 5-Asset Analysis
**Date:** June 2026
**Period analysed:** January 2019 - January 2024
**Author:** JUSTOH-CPP


---

## Thesis
Across 5 major asset classes, AAPL delivered the strongest risk-adjusted 
return (Sharpe 0.95) over the 2019-2024 period, outperforming the S&P 500 
benchmark on every risk-adjusted metric. TLT was the worst performer due 
to the 2022 rate hiking cycle destroying long-duration bond prices.

---

## Key Metrics

| Asset   | Ann Return | Volatility | Sharpe | Max Drawdown |
|---------|-----------|------------|--------|--------------|
| AAPL    | 32.59%    | 32.25%     | 0.95   | -35.18%      |
| SPY     | 14.50%    | 21.08%     | 0.59   | -35.75%      |
| BTC-USD | 33.11%    | 56.25%     | 0.55   | -83.72%      |
| GLD     | 9.11%     | 14.95%     | 0.48   | -23.89%      |
| TLT     | -2.02%    | 17.63%     | -0.23  | -51.11%      |

---

## Distribution Findings
- No asset follows a normal distribution — all rejected by statistical tests
- BTC has the most extreme fat tails (kurtosis 19.05) — extreme daily moves 
  happen far more often than any model assuming normality would predict
- GLD has the most normal-like distribution (kurtosis 3.07) — most 
  predictable of the 5 assets
- SPY shows strong negative skew (-0.82) — losses are more extreme than gains

---

## Current Signals (as of end 2023)
- SPY RSI at 70.54 — overbought, potential mean reversion sell signal
- GLD is the only asset with a BULLISH MACD signal
- AAPL and BTC both neutral on RSI around 51

---

## Risks
- AAPL outperformance may not persist — single stock concentration risk
- BTC -83% drawdown shows crypto requires extreme risk tolerance
- TLT analysis covers a uniquely bad rate environment — bonds may recover
- All metrics are backward-looking — past performance does not guarantee future results

---

## Conclusion
For a risk-aware quantitative strategy, AAPL and SPY offer the best 
risk-adjusted entry points. BTC offers high returns but requires position 
sizing at 25-30% of normal size given its tail risk profile. TLT is 
uninvestable until the rate environment stabilises.